clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -237,12 +239,18 @@ public void close() {
      * @param send The request to send
      */
     public void send(Send send) {
-        KafkaChannel channel = channelOrFail(send.destination());
-        try {
-            channel.setSend(send);
-        } catch (CancelledKeyException e) {
-            this.failedSends.add(send.destination());
-            close(channel);
+        String connectionId = send.destination();
+        KafkaChannel closingChannel = closingChannels.remove(connectionId);
+        if (closingChannel != null)
+            doClose(closingChannel, true);
```

Comment:
The problem is `clear()` is iterating through `closingChannels` so the code that removes the channel from `closingChannels` needs to use the iterator in `clear`. So the common code is just calling `doClose`. Not sure if that is worth is a separate method.